### PR TITLE
chore(v2): Fix linter warnings

### PR DESCRIPTION
**Pull request by SamChou19815**
## Motivation

Help reduce number of linter warnings. .

223 warnings to 148 warnings.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I avoided changes that can potentially cause runtime behavior changes. All the changes here are almost all type-only. Since it passes the type checker and linter, it shouldn't affect any runtime behavior.

## Related PRs

N/A

### CODE REVIEW

File: packages/docusaurus-theme-common/src/utils/docsPreferredVersion/useDocsPreferredVersion.ts [hunk 1]
```diff
@@ -15,7 +15,10 @@ import {DEFAULT_PLUGIN_ID} from '@docusaurus/constants';
 // Note, the preferredVersion attribute will always be null before mount
 export function useDocsPreferredVersion(
   pluginId: string | undefined = DEFAULT_PLUGIN_ID,
-) {
+): {
+  readonly preferredVersion: any;
+  readonly savePreferredVersionName: (versionName: string) => void;
+} {
```
Comment (slorber) [inline packages/docusaurus-theme-common/src/utils/docsPreferredVersion/useDocsPreferredVersion.ts:21]: do we really want to have explicit returns everywhere?

I find this a bit annoying personally and like to leverage inference in some cases

File: packages/docusaurus-theme-common/src/utils/useThemeConfig.ts [hunk 2]
```diff
@@ -32,11 +34,39 @@ export type Navbar = {
   logo?: NavbarLogo;
 };
 
+export type ColorModeConfig = {
```
Comment (slorber) [inline packages/docusaurus-theme-common/src/utils/useThemeConfig.ts:37]: FYI @armano2 this is likely to conflict with your work here: https://github.com/facebook/docusaurus/pull/4433

File: packages/docusaurus/src/client/exports/Link.tsx [hunk 3]
```diff
@@ -113,22 +116,22 @@ function Link({
     if (IOSupported && ref && isInternal) {
       // If IO supported and element reference found, setup Observer functionality.
       handleIntersection(ref, () => {
-        window.docusaurus.prefetch(targetLink);
+        window.docusaurus.prefetch(targetLink || '');
```
Comment (slorber) [inline packages/docusaurus/src/client/exports/Link.tsx:119]: Should we prefetch anything if there is no targetLink ?

File: packages/docusaurus/src/client/exports/Link.tsx [hunk 4]
```diff
@@ -113,22 +116,22 @@ function Link({
...
       });
     }
   };
 
   const onMouseEnter = () => {
     if (!preloaded.current) {
-      window.docusaurus.preload(targetLink);
+      window.docusaurus.preload(targetLink || '');
```
Comment (slorber) [inline packages/docusaurus/src/client/exports/Link.tsx:126]: same

File: packages/docusaurus/src/client/exports/Link.tsx [hunk 5]
```diff
@@ -113,22 +116,22 @@ function Link({
...
       preloaded.current = true;
     }
   };
 
   useEffect(() => {
     // If IO is not supported. We prefetch by default (only once).
     if (!IOSupported && isInternal) {
-      window.docusaurus.prefetch(targetLink);
+      window.docusaurus.prefetch(targetLink || '');
```
Comment (slorber) [inline packages/docusaurus/src/client/exports/Link.tsx:134]: same

File: packages/docusaurus-theme-common/src/utils/usePluralForm.ts [hunk 6]
```diff
@@ -107,7 +107,9 @@ function selectPluralMessage(
   }
 }
 
-export function usePluralForm() {
+export function usePluralForm(): {
+  selectMessage: (count: number, pluralMessages: string) => string;
+} {
```
Comment (slorber) [inline packages/docusaurus-theme-common/src/utils/usePluralForm.ts:112]: also prefer inference here

File: packages/docusaurus-theme-common/src/utils/useAlternatePageUtils.ts [hunk 7]
```diff
@@ -11,7 +11,15 @@ import {useLocation} from '@docusaurus/router';
...
-export function useAlternatePageUtils() {
+export function useAlternatePageUtils(): {
+  createUrl: ({
+    locale,
+    fullyQualified,
+  }: {
+    locale: string;
+    fullyQualified: boolean;
+  }) => string;
+} {
```
Comment (slorber) [inline packages/docusaurus-theme-common/src/utils/useAlternatePageUtils.ts:22]: 😅 I'd prefer just using inference here

**Review by slorber — CHANGES_REQUESTED** [review]
thanks, almost good

just wondering if it's not better to use TS inference for some return types? 

Particularly on hooks that are not public API, I find declaring full fn signature type creates a lot of boilerplate

**Comment by slorber** [comment]
thanks :)

File: packages/docusaurus-theme-common/src/utils/useThemeConfig.ts (hunk 2 above)
Comment (armano2) [reply packages/docusaurus-theme-common/src/utils/useThemeConfig.ts:37]: no worries, i have to redo this pr anyway, and i'm going to split it to smaller pieces

**Pull request merged**